libyul/optimiser/ReasoningBasedSimplifier.cpp

```diff
@@ -0,0 +1,194 @@
+
+	m_solver->push();
+	m_solver->addAssertion(condition != int2bv(0));
+
+	ASTModifier::operator()(_if.body);
+
+	m_solver->pop();
+}
+
+ReasoningBasedSimplifier::ReasoningBasedSimplifier(
+	Dialect const& _dialect,
+	set<YulString> const& _ssaVariables
+):
+	m_dialect(_dialect),
+	m_ssaVariables(_ssaVariables)
+{
+	m_solver = make_unique<smtutil::Z3Interface>();
+}
+
+smtutil::Expression ReasoningBasedSimplifier::encodeExpression(Expression const& _expression)
+{
+	return std::visit(GenericVisitor{
+		[&](FunctionCall const& _functionCall)
+		{
+			if (auto const* dialect = dynamic_cast<EVMDialect const*>(&m_dialect))
+				if (auto const* builtin = dialect->builtin(_functionCall.functionName.name))
+					if (builtin->instruction)
+						return encodeBuiltin(*builtin->instruction, _functionCall.arguments);
+			return newVariable();
+		},
+		[&](Identifier const& _identifier)
+		{
+			if (
+				m_ssaVariables.count(_identifier.name) &&
+				m_variables.count(_identifier.name)
+			)
+				return m_variables.at(_identifier.name);
+			else
+				return newVariable();
+		},
+		[&](Literal const& _literal)
+		{
+			return int2bv(smtutil::Expression(valueOfLiteral(_literal)));
+		}
+	}, _expression);
+}
+
+smtutil::Expression ReasoningBasedSimplifier::encodeBuiltin(
+	evmasm::Instruction _instruction,
+	vector<Expression> const& _arguments
+)
+{
+	vector<smtutil::Expression> arguments = applyMap(
+		_arguments,
+		[this](Expression const& _expr) { return encodeExpression(_expr); }
+	);
+	switch (_instruction)
+	{
+	case evmasm::Instruction::ADD:
+		return (arguments.at(0) + arguments.at(1));
+	case evmasm::Instruction::MUL:
+		return (arguments.at(0) * arguments.at(1));
+	case evmasm::Instruction::SUB:
+		return (arguments.at(0) - arguments.at(1));
+	case evmasm::Instruction::DIV:
+		return smtutil::Expression::ite(
+			arguments.at(1) == int2bv(0),
+			int2bv(0),
+			arguments.at(0) / arguments.at(1)
+		);
+	// TODO SDIV
+	case evmasm::Instruction::MOD:
+		return smtutil::Expression::ite(
+			arguments.at(1) == int2bv(0),
+			int2bv(0),
+			arguments.at(0) % arguments.at(1)
+		);
+	// TODO SMOD
+	case evmasm::Instruction::LT:
+		return smtutil::Expression::ite(arguments.at(0) < arguments.at(1), int2bv(1), int2bv(0));
+	case evmasm::Instruction::GT:
+		return smtutil::Expression::ite(arguments.at(0) > arguments.at(1), int2bv(1), int2bv(0));
+	case evmasm::Instruction::EQ:
+		return smtutil::Expression::ite(arguments.at(0) == arguments.at(1), int2bv(1), int2bv(0));
+	case evmasm::Instruction::ISZERO:
+		return smtutil::Expression::ite(arguments.at(0) == int2bv(0), int2bv(1), int2bv(0));
+	case evmasm::Instruction::AND:
+		return arguments.at(0) & arguments.at(1);
+	case evmasm::Instruction::OR:
+		return arguments.at(0) | arguments.at(1);
+	case evmasm::Instruction::NOT:
+		return ~arguments.at(0);
+	case evmasm::Instruction::SHL:
+		return arguments.at(0) << arguments.at(1);
+	default:
```

Comment:
Also missing XOR, SHR (SAR -- though this is tricky and not used frequently, i.e. not worth the hassle).

Reply:
https://github.com/Z3Prover/z3/blob/de9bc930e9b3d77f971bd2059359bdd8bd5b89e4/src/api/c%2B%2B/z3%2B%2B.h#L1869

Apparently has both arithmetic and logic right shifts, thought not fully if it is compatible with our interpretation.

Reply:
Added both in the other PR.